### DISCUSSION TITLE
How to efficiently get top K results with a filtering condition involving lots of OR

### DISCUSSION THREAD

**Discussion by jiazhouplusai** [milvus-io/milvus#35202] [Q&A and General discussion]
I am trying to use milvus for image similarity search involving 100 million+ embeddings. We have a different column-oriented distributed db table outside of milvus to annotate hundreds of metas for a given image like latitude, longitude, # of pedestrians, etc.

When we do search, we want to find top K images which also satisfying metas filtering condition. So we first make a call against above meta table to get all image ids satisfying specific meta conditions. (e.g., latitude > x and longitude < y and # of pedestrians > 0) We then want to find top K results using the text query embedding plus a filtering condition potentially involving lots of images ids like "image_id == id1 OR image_id == id2 OR ... OR image_id == id1000000". 

So far the latency for this top K + id filtering query against milvus is large and we are trying to optimize it.

One way is to also store all metas inside milvus so we don't have to filter by tons of OR for selected ids. However, there are lots of metas and they are already being used by other use cases so at this stage we don't want to maintain a separate copy inside milvus along with image embeddings.

On the other hand, inside milvus, we can also apply indexing based on image_id. After that, we can partition the data based on image_id into N partitions, on application level split filtering id sets into N batches (one for a different partition), then split to make N queries for top K inside each partition and finally, merge the results for final top K. This sounds something that may help but requires non-trival logic on application level.

We want to understand whether there is similar case like us already and what are some good ways to speed up the query in such scenario? Thanks in advance for your time/help!

**Comment by yhmo** [comment]
Sounds like you are performing 2-steps query:
1. query ids meet the conditions:
```
results = collection.query(expr="latitude > x and longitude < y and pedestrians > 0", output_fields=["id"])
ids = [res["id"] for res in results]
print(ids)
```

2. ANN search in the embeddings of the ids:
```
results = collection.search(data=[text query embedding],
                            limit=topk,
                            expr=f"id in {ids}",
                            anns_field="xxxx",
                            param={},
                            )
```

Why not combine the two steps into one step?
```
results = collection.search(data=[text query embedding],
                            limit=topk,
                            expr="latitude > x and longitude < y and pedestrians > 0",
                            anns_field="xxxx",
                            param={},
                            )
```

**Comment by jiazhouplusai** [comment]
Thanks @yhmo for taking a look at our use case! We use a **different** column-oriented distributed db table **outside** of milvus to annotate hundreds of metas for a given image like latitude, longitude, # of pedestrians, etc.  One way is to also store all metas inside milvus so we don't have to filter by tons of OR for selected ids. However, there are lots of metas and they are already being used by other use cases so at this stage we don't want to maintain a separate copy inside milvus along with image embeddings.

**Reply by yhmo** [reply]
Let's say we have a collection with 100 million+ embeddings in milvus. 
And there are hundreds of meta columns for each embedding. The meta data is stored in an outside database.

Now we want to conduct a ANN search with the condition "latitude > x and longitude < y and pedestrians > 0". Assume there are 100K embeddings matching this condition. With the 2-steps query, finally we get topK embeddings from the 100K embeddings.

I have a question: do you need meta data of the 100K embeddings, or the topK embeddings?

================================================================================================
If you only need meta data of the topK embeddings, you can use another 2-steps query to avoid long-length filter expression.
Step 1, do ANN search with the condition "latitude > x and longitude < y and pedestrians > 0"
```
results = collection.search(data=[text query embedding],
                            limit=topk,
                            expr="latitude > x and longitude < y and pedestrians > 0",
                            anns_field="xxxx",
                            param={},
                            )
```
The results contain topK items with ID-distance pairs.

Step 2, use the topK IDs to query meta data from the outside database.

================================================================================================
If you need meta data of all the 100K embedding, I think the condition "image_id == id1 OR image_id == id2 OR ... OR image_id == id1000000" is equivalent to "latitude > x and longitude < y and pedestrians > 0".
Step 1, query IDs with the condition "latitude > x and longitude < y and pedestrians > 0"
`results = collection.query(expr="latitude > x and longitude < y and pedestrians > 0", output_fields=["id"])`

Step 2, use the IDs to fetch meta data from the outside database

Step 3, since the condition "image_id == id1 OR image_id == id2 OR ... OR image_id == id1000000" is equivalent to "latitude > x and longitude < y and pedestrians > 0", we can do ANN search with the condition "latitude > x and longitude < y and pedestrians > 0"
```
results = collection.search(data=[text query embedding],
                            limit=topk,
                            expr="latitude > x and longitude < y and pedestrians > 0",
                            anns_field="xxxx",
                            param={},
                            )
```

================================================================================================
I think both solutions can avoid the long-length expression and get the correct result, you can have a try.

**Reply by jiazhouplusai** [reply]
Thanks @yhmo ! In my case, milvus doesn't store any meta information like latitude/longitude/pedestrians and they are stored outside of milvus. In the end, I need to get top K images satisfying meta condition filtering and also only need to get corresponding metas for these top K results from the outside database (which is fast so far)

**Comment by xiaofan-luan** [comment]
OR operation is efficient, but it is not recommend to OR many conditions.
Simple math:
You have 10m corpus, you need 32 bit to indicate each PK, with 100K IN or OR operation, the condition expression itself is more than 3.2MB. There is no way to parse this expr very quick.

**Comment by xiaofan-luan** [comment]
My suggestion is either to put this data in milvus or use post filtering.

With reasonable filtering ratio (like 20%), if you want topk = 10, you can trigger a search with topk = 100 and filter the 10 result match.

You can also use range search  and iterate through the result

**Comment by jiazhouplusai** [comment]
Thanks @xiaofan-luan for your suggestions! I will explore the post filtering approach or even a hybrid approach involving both post filtering and direct filtering depending on the percentage/number of images satisfying meta condition.
